src/components/tl/sharp/tl_sharp_coll.c

```diff
@@ -82,21 +82,32 @@ static inline ucc_status_t ucc_tl_sharp_status_to_ucc(int status)
 static ucc_tl_sharp_reg_t ucc_tl_sharp_reg_null = { .mr = NULL };
 
 static ucc_status_t
-ucc_tl_sharp_mem_register(ucc_tl_sharp_context_t *ctx, void *addr,
+ucc_tl_sharp_mem_register(ucc_tl_sharp_context_t *ctx,
+                          ucc_tl_sharp_team_t *team, void *addr,
                           size_t length, ucc_tl_sharp_reg_t **reg)
 {
     ucc_rcache_region_t          *rregion;
     ucc_tl_sharp_rcache_region_t *region;
     ucc_status_t                  status;
     ucc_tl_sharp_reg_t           *r;
+    ucc_rcache_t                 *rcache;
+    struct sharp_coll_context    *sharp_ctx;
 
     if (length < ctx->cfg.reg_threshold) {
         *reg = &ucc_tl_sharp_reg_null;
         return UCC_OK;
     }
 
-    if (ctx->rcache) {
-        status = ucc_rcache_get(ctx->rcache, (void *)addr, length, NULL,
+    if (ctx->sharp_context  != NULL) {
+        sharp_ctx = ctx->sharp_context;
```

Comment:
maybe always set team->sharp_ctx during team creation. If CTX_PER_TEAM is not set then the ptr will point to global context. Then you don't need these branches, and the ucc_tl_sharp_mem_register can just take team